src/charm.py

```diff
@@ -93,36 +87,42 @@ def _on_leader_elected(self, _) -> None:
         data.setdefault("replication-password", self._new_password())
 
     def _on_start(self, event) -> None:
-        password = self._get_postgres_password()
-        # If the leader was elected and it generated a superuser password for the all the units,
-        # the cluster can be bootstrapped in each unit.
-        if password is not None:
-            try:
-                self._cluster.bootstrap_cluster(password)
-            except ClusterAlreadyRunningError:
-                logging.error("there is already a running cluster")
-                self.unit.status = BlockedStatus("there is already a running cluster")
-            except ClusterCreateError as e:
-                logging.error("failed to create cluster")
-                self.unit.status = BlockedStatus(f"failed to create cluster with error {e}")
-            except (ClusterNotRunningError, ClusterStartError) as e:
-                logging.error("failed to start cluster")
-                self.unit.status = BlockedStatus(f"failed to start cluster with error {e}")
-            except subprocess.CalledProcessError as e:
-                logging.error("failed to bootstrap cluster")
-                self.unit.status = BlockedStatus(f"failed to bootstrap cluster with error {e}")
-            else:
+        """Handle the start event."""
+        # Doesn't try to bootstrap the cluster if it's in a blocked state
+        # caused, for example, because a failed installation of packages.
+        if self._has_blocked_status:
+            return
```

Comment:
would `return`ing or `defer`ing be right thing to do here? i think `return`, but want to confirm the logic - once in a blocked status, the human operator will resolve all issues, run all actions not run and  then run `upgrade-charm`. is that right?

Reply:
I was reading about BlockedStatus and the `juju resolved` command, and I think if certain hooks run while the charm is blocked, it makes sense to raise an error instead. My reasoning is that a BlockedStatus should be raised only when a user is required to fix the charm. Once they finish fixing whatever caused the blockage, they can run juju resolved to re-run failed hooks, so hooks such as `on_start` should fail when the charm is blocked to make use of this command. If you go this route, it is worth documenting that this is the intended behaviour on BlockedStatus.

Incidentally, I don't think a user should need to run upgrade-charm when fixing a blockage, since they would likely be downloading the charm again.

Reply:
Thanks for the explanation, your approach makes more sense Marcelo :)